src/Mod/Sketcher/App/SketchObject.cpp

```diff
@@ -2970,160 +2978,250 @@ int SketchObject::split(int GeoId, const Base::Vector3d &point)
 
+    auto createGeosFromPeriodic =
+        [&](const Part::GeomCurve* curve,
+            auto getCurveWithLimitParams,
+            auto createAndTransferConstraints) {
+            // find split point
+            curve->closestParameter(point, splitParam);
+            double period = curve->getLastParameter() - curve->getFirstParameter();
+            startParam = splitParam;
+            endParam = splitParam + period;
+
+            // create new arc and restrict it
+            auto newCurve = getCurveWithLimitParams(curve, startParam, endParam);
+            int newId(GeoEnum::GeoUndef);
+            newId = addGeometry(std::move(newCurve)); // after here newCurve is a shell
+            if (newId >= 0) {
+                newIds.push_back(newId);
+                setConstruction(newId, GeometryFacade::getConstruction(curve));
+                exposeInternalGeometry(newId);
 
-    bool ok = false;
-    if (geo->getTypeId() == Part::GeomLineSegment::getClassTypeId()) {
-        const Part::GeomLineSegment *lineSegm = static_cast<const Part::GeomLineSegment *>(geo);
-
-        startPoint = lineSegm->getStartPoint();
-        endPoint = lineSegm->getEndPoint();
-        splitPoint = point.Perpendicular(startPoint, endPoint - startPoint);
-        if ((endPoint - splitPoint).Length() > (splitPoint - startPoint).Length()) {
-            longestPart = 1;
-        }
-
-        Part::GeomLineSegment *newLine = static_cast<Part::GeomLineSegment *>(lineSegm->copy());
-        newGeometries.push_back(newLine);
+                // transfer any constraints
+                createAndTransferConstraints(GeoId, newId);
+                return true;
+            }
 
-        newLine->setPoints(startPoint, splitPoint);
-        int newId = addGeometry(newLine);
-        if (newId >= 0) {
-            newIds.push_back(newId);
-            setConstruction(newId, GeometryFacade::getConstruction(geo));
+            return false;
+        };
 
-            newLine = static_cast<Part::GeomLineSegment*>(lineSegm->copy());
-            newGeometries.push_back(newLine);
+    auto createGeosFromNonPeriodic =
+        [&](const Part::GeomBoundedCurve* curve,
+            auto getCurveWithLimitParams,
+            auto createAndTransferConstraints) {
+            startPoint = curve->getStartPoint();
+            endPoint = curve->getEndPoint();
+
+            // find split point
+            curve->closestParameter(point, splitParam);
+            startParam = curve->getFirstParameter();
+            endParam = curve->getLastParameter();
+            // TODO: Using parameter difference as a poor substitute of length.
+            // Computing length of an arc of a generic conic would be expensive.
+            if (endParam - splitParam < Precision::PConfusion() ||
+                splitParam - startParam < Precision::PConfusion()) {
+                THROWM(ValueError, "Split point is at one of the end points of the curve.");
+                return false;

```

Comment:
There are some considerations here:
1. The SketchObject split function returns an **int**, with an error code being a negative error code. This lambda returns a **boolean**, so it cannot convey an error code. An exception may be used to convey a specific message (THROWM). In this case, we are throwing a c++ exception (via THROWM), so ``` return false``` is unreachable (the execution path will never arrive there). This should be removed.
 
2. There is a choice to be made, for a given error, that the function may be left via c++ exception or via error code, but it cannot be both at the same time (i.e. for the same specific error). 

3. If error code is favoured and a function that is called may return an exception (as in the case of trim() ), then the exceptions are caught at SketchObject level and the corresponding error code is returned.
 
4. If exception is favoured, then the caller of this function shall catch it. When the caller is the Python binding (SketchObjectPyImp), it must catch it before leaving c++ land towards Python land. 

I am not sure myself why exactly the C++ exception must be converted to a Python exception. I wanted to have a conversation with @wwmayer about it, but suddenly I was not available and then I forgot. This can be a good occasion to ask him about the limitations in place and the underlying reasons.

In practical terms, because in SketchObject some functions just return error codes, but other functions may leave by exception (at least for some errors), in SketchObjectPyImp.cpp there are different approaches to this.

**Way A)** pure error code land:
```
    int err=this->getSketchObjectPtr()->setDatum(Index, Datum);
    if (err) {
        std::stringstream str;
        if (err == -1)
            str << "Invalid constraint index: " << Index;
        else if (err == -3)
            str << "Cannot set the datum because the sketch contains conflicting constraints";
        else if (err == -2)
            str << "Datum " << (const char*)Quantity.getUserString().toUtf8() << " for the constraint with index " << Index << " is invalid";
        else if (err == -4)
            str << "Negative datum values are not valid for the constraint with index " << Index;
        else if (err == -5)
            str << "Zero is not a valid datum for the constraint with index " << Index;
        else if (err == -6)
            str << "Cannot set the datum because of invalid geometry";
        else
            str << "Unexpected problem at setting datum " << (const char*)Quantity.getUserString().toUtf8() << " for the constraint with index " << Index;
        PyErr_SetString(PyExc_ValueError, str.str().c_str());
        return nullptr;
```
So handle the error codes and create Python exceptions (PyErr_SetString). No c++ exception reach here (if any, they were caught at SketchObject level and converted into error codes).

When we do this (PyErr_SetString), and we called the function from SketcherGui, we end up with a c++ exception at the caller (when Python returns control to the c++ code invoking the Python command, the python exception generates a c++ exception):

```
            try {
                Gui::Command::openCommand(QT_TRANSLATE_NOOP("Command", "Split edge"));
                Gui::cmdAppObjectArgs(sketchgui->getObject(), "split(%d,App.Vector(%f,%f,0))",
                                      GeoId, onSketchPos.x, onSketchPos.y);
                Gui::Command::commitCommand();
                tryAutoRecompute(static_cast<Sketcher::SketchObject *>(sketchgui->getObject()));
            }
            catch (const Base::Exception& e) {
                Base::Console().Error("Failed to split edge: %s\n", e.what());
                Gui::Command::abortCommand();
            }
```
So, the transaction guarantees are enforced here and any change to properties reverted. The message also contains the string that was put with ```PyErr_SetString```.

**WAY B)** Exiting SketchObject function via exception or error code (for some specific causes one, or other specific causes the other)

Then we need to catch this c++ exception and convert it into a Python exception.

```        try {
            int ret = this->getSketchObjectPtr()->addCopy(geoIdList, vect, false, Base::asBoolean(clone)) + 1;

            if(ret == -1)
                throw Py::TypeError("Copy operation unsuccessful!");

            std::size_t numGeo = geoIdList.size();
            Py::Tuple tuple(numGeo);
            for (std::size_t i=0; i<numGeo; ++i) {
                int geoId = ret - int(numGeo - i);
                tuple.setItem(i, Py::Long(geoId));
            }

            return Py::new_reference_to(tuple);
        }
        catch(const Base::ValueError & e) {
            throw Py::ValueError(e.getMessage());
        }```

Here I am looking into the catch block, which throws a Python exception. I think this is equivalent to the PyErr_SetString, but @wwmayer knows better and may confirm or deny this assertion or point out to differences.

So in this case, error codes and c++ exceptions, end up becoming Python exceptions. If the caller of the Python command was c++, then the caller gets a c++ exception corresponding to the Python counterpart, which shall be used to provide the transactional guarantee of a command.

N.B.: Often in this  post I assume that the caller of the SketchObject c++ function is Python, as it is the cases here. It needs not be, it can be another c++ function, but that is a different discussion.

Reply:
OK I see the issue here. I had just followed snippets elsewhere in `SketchObject.cpp` etc. I have no preference from my end.

Reply:
I think it makes sense in this case to exit by THROWM, then catch the exception in SketchObjectPyImp, and either set an error with PyErr_SetString, or raise a Py::ValueError exception. I think that is what is left from your side.

In parallel I am tweaking the mechanism for notification of the parabola migration. I hope to have final code on my side this week.

Then with the two this PR can be merged. I find it is a very nice functionality.

Reply:
Are you sure we should do that? I don't see try-catch being used a lot in `SketchObjectPyImp.cpp`.